src/spark_rapids_ml/core.py

```diff
@@ -438,15 +454,20 @@ def _train_udf(pdf_iter: Iterator[pd.DataFrame]) -> pd.DataFrame:
             if enable_nccl:
                 context.barrier()
 
-            if context.partitionId() == 0:
+            if return_model == True:
+                if context.partitionId() == 0:
+                    yield pd.DataFrame(data=result)
+            else:
                 yield pd.DataFrame(data=result)
 
-        ret = (
-            dataset.mapInPandas(_train_udf, schema=self._out_schema())  # type: ignore
-            .rdd.barrier()
-            .mapPartitions(lambda x: x)
-            .collect()[0]
-        )
+        barrier_rdd = dataset.mapInPandas(
+            _train_udf, schema=self._out_schema()
+        ).rdd.barrier()  # type: ignore
+
+        if return_model == False:
+            return barrier_rdd.mapPartitions(lambda x: x)
+
+        ret = barrier_rdd.mapPartitions(lambda x: x).collect()[0]
```

Comment:
I'd move `mapPartitions(lambda x: x)` into the common code above, since it's duplicated otherwise.

Reply:
Done.